File: app/src/main/java/br/pedroso/upcomingmovies/moviedetails/MovieDetailsPresenter.java
package br.pedroso.upcomingmovies.moviedetails;

import java.util.List;

import javax.inject.Inject;

import br.pedroso.upcomingmovies.domain.Movie;
import br.pedroso.upcomingmovies.domain.MoviesRepository;
import br.pedroso.upcomingmovies.moviedetails.MovieDetailsView;
import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers;
import io.reactivex.rxjava3.disposables.CompositeDisposable;
import io.reactivex.rxjava3.disposables.Disposable;

public class MovieDetailsPresenter {

    private final MovieDetailsView view;

    private final MoviesRepository moviesRepository;

    private final CompositeDisposable disposables = new CompositeDisposable();

    @Inject
    public MovieDetailsPresenter(MovieDetailsView view, MoviesRepository moviesRepository) {
        this.view = view;
        this.moviesRepository = moviesRepository;
    }

    public void loadMovieDetails(int movieId) {
        Disposable disposable = moviesRepository.getMovieDetails(movieId)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(this::displayMovieDetails, this::displayLoadingErrorMessage);

        disposables.add(disposable);

        loadSimilarMovies(movieId);
    }

    private void displayLoadingErrorMessage(Throwable throwable) {
        // TODO
    }

    private void loadSimilarMovies(int movieId) {
        Disposable disposable = moviesRepository.listSimilarMovies(movieId)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(this::displaySimilarMovies, error -> view.hideSimilarMoviesPanel());

        disposables.add(disposable);
    }

    private void displayMovieDetails(Movie movie) {
        view.renderMovieDetails(movie);
    }

    private void displaySimilarMovies(List<Movie> movies) {
        if (!movies.isEmpty()) {
            view.displaySimilarMoviesPanel();
            view.renderSimilarMovies(movies);
        } else {
            view.hideSimilarMoviesPanel();
        }
    }

    public void pause() {
        disposables.clear();
    }
}

File: app/src/main/kotlin/br/pedroso/upcomingmovies/moviedetails/MovieDetailsPresenter.kt
package br.pedroso.upcomingmovies.moviedetails

import br.pedroso.upcomingmovies.domain.Movie
import br.pedroso.upcomingmovies.domain.MoviesRepository
import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers
import io.reactivex.rxjava3.disposables.CompositeDisposable
import javax.inject.Inject

class MovieDetailsPresenter @Inject constructor(
    private val view: MovieDetailsView,
    private val moviesRepository: MoviesRepository
) {
    private val disposables = CompositeDisposable()

    fun loadMovieDetails(movieId: Int) {
        val disposable = moviesRepository.getMovieDetails(movieId)
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe(
                { movie: Movie -> this.displayMovieDetails(movie) },
                { throwable: Throwable -> this.displayLoadingErrorMessage(throwable) })

        disposables.add(disposable)

        loadSimilarMovies(movieId)
    }

    private fun displayLoadingErrorMessage(throwable: Throwable) {
        // TODO
    }

    private fun loadSimilarMovies(movieId: Int) {
        val disposable = moviesRepository.listSimilarMovies(movieId)
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe(
                { movies: List<Movie> -> this.displaySimilarMovies(movies) },
                { error: Throwable? -> view.hideSimilarMoviesPanel() })

        disposables.add(disposable)
    }

    private fun displayMovieDetails(movie: Movie) {
        view.renderMovieDetails(movie)
    }

    private fun displaySimilarMovies(movies: List<Movie>) {
        if (movies.isNotEmpty()) {
            view.displaySimilarMoviesPanel()
            view.renderSimilarMovies(movies)
        } else {
            view.hideSimilarMoviesPanel()
        }
    }

    fun pause() {
        disposables.clear()
    }

    fun clickedOnSimilarMovie(movie: Movie) {
        view.startMovieDetailsActivity(movie.id)
    }
}
